### PR TITLE
Introducing the maximum size preview

### PR DESCRIPTION
**Pull request by oparoz**
#### Fixes
- Always store the preview returned by Imagemagick before resizing it. https://github.com/owncloud/core/issues/12465
#### The problem

You've just uploaded a 100MB TIFF file that you need to share with a restricted audience. Here is what currently happens once the file has been uploaded.
- A 36x36 thumbnail is generated in 1 minute
- Go to the Gallery, an album thumbnail might be generated in 1 minute
- Go inside the album, another thumbnail will be generated in 1 minute
- Click on the picture to get a full screen preview. 1 minute
- Go back to the Files app and share the file
- Client 1 visits the link. 1 minute to get the preview
- Client 2 with a different screen resolution visits the link. 1 minute again

You're starting to get it. Every time a slightly different preview size is asked, a full conversion of that 100MB TIFF to PNG is performed. This is highly inefficient.
#### Proposed solution
- Enforce maximum preview dimensions in the config. They're currently set to null.
- The first time we're asked to generate a preview we'll generate one of the maximum dimension indicated in the configuration and all future resizing requests will be done on that preview in order to not waste time converting the same large file over and over

Default size is 2048x2048
The maximum size preview is keeping the aspect ratio of the original and is not upscaled.
It's is immediately stored in the cache with a special `max` tag and size identifiers based on its real size.

Every time a preview of a different size is asked, we store it in the cache. This also speeds up operation at the cost of disk space.

This make a huge difference in libraries holding several large files, for people using apps such as the Gallery.

### CODE REVIEW

**Comment by karlitschek** [comment]
@georgehrke

**Comment by DeepDiver1975** [comment]
@oparoz thanks a lot! really awesome! Let schedule this for 8.1

**Comment by georgehrke** [comment]
:+1: for the proposed solution! Thanks a lot.

Will review the code asap

**Comment by DeepDiver1975** [comment]
> Will review the code asap

@georgehrke after OC8 has been released please - THX

File: lib/private/preview.php [hunk 1]
```diff
@@ -680,11 +698,128 @@ private function resizeAndCrop() {
...
+			}
+
+			// This stores our large preview so that it can be used in subsequent resizing requests
+			$this->storeMaxPreview($previewPath);
+
+			break;
+		}
+
+		// The providers have been kind enough to give us a preview
+		if (!is_null($preview)) {
```
Comment (oparoz) [inline lib/private/preview.php]: Use `if (preview)` instead as provider's response can be boolean

**Pull request closed without merging**